src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Variables.cs

```diff
@@ -0,0 +1,373 @@
+
+            internal int RootSlot(int slot)
+            {
+                while (true)
+                {
+                    int containingSlot = this[slot].ContainingSlot;
+                    if (containingSlot == 0)
+                    {
+                        return slot;
+                    }
+                    else
+                    {
+                        slot = containingSlot;
+                    }
+                }
+            }
+
+            internal bool TryGetValue(VariableIdentifier identifier, out int slot)
+            {
+                var variables = GetVariablesForVariable(identifier);
+                return variables.TryGetValueInternal(identifier, out slot);
+            }
+
+            private bool TryGetValueInternal(VariableIdentifier identifier, out int slot)
+            {
+                if (_variableSlot.TryGetValue(identifier, out int index))
+                {
+                    slot = ConstructSlot(Id, index);
+                    return true;
+                }
+                slot = -1;
+                return false;
+            }
+
+            internal int Add(VariableIdentifier identifier)
+            {
+                if (Count > IndexMask)
+                {
+                    return -1;
+                }
+                if (GetSlotDepth(identifier.ContainingSlot) >= MaxSlotDepth)
+                {
+                    return -1;
+                }
+                var variables = GetVariablesForVariable(identifier);
+                return variables.AddInternal(identifier);
+            }
+
+            private int AddInternal(VariableIdentifier identifier)
+            {
+                int index = Count;
+                _variableSlot.Add(identifier, index);
+                while (index >= _variableBySlot.Count)
+                {
+                    _variableBySlot.Add(default);
+                }
+                _variableBySlot[index] = identifier;
+                return ConstructSlot(Id, index);
+            }
+
+            internal bool TryGetType(Symbol symbol, out TypeWithAnnotations type)
+            {
+                var variables = GetVariablesContainingVariable(symbol);
+                return variables._variableTypes.TryGetValue(symbol, out type);
+            }
+
+            internal void SetType(Symbol symbol, TypeWithAnnotations type)
+            {
+                var variables = GetVariablesContainingVariable(symbol);
+                Debug.Assert((object)variables == this);
+                variables._variableTypes[symbol] = type;
+            }
+
+            internal VariableIdentifier this[int slot]
+            {
+                get
+                {
+                    (int id, int index) = DeconstructSlot(slot);
+                    var variables = GetVariablesForId(id);
+                    return variables!._variableBySlot[index];
+                }
+            }
+
+            internal int Count => _variableBySlot.Count;
```

Comment:
>internal int Count [](start = 12, length = 18)

Is there a good reason to expose this count to other components?

Reply:
Perhaps the name should be changed to avoid confusion around this property.

---
In reply to: [561462447](https://github.com/dotnet/roslyn/pull/50417#discussion_r561462447) [](ancestors = 561462447)